btcpayserver: optional altcoin support

nixbitcoin opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change


###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [x] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Relase notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

In use on https://github.com/fort-nix/nix-bitcoin/pull/369

nixbitcoin committed:
btcpayserver: optional altcoin support

prusnak commented:
@GrahamcOfBorg build btcpayserver

prusnak commented on pkgs/applications/blockchains/btcpayserver/default.nix:
```diff
@@ -34,8 +36,8 @@ stdenv.mkDerivation rec {
 
-    dotnet restore --source $TMP/nuget BTCPayServer/BTCPayServer.csproj
-    dotnet publish --no-restore --output $out/share/$pname -c Release BTCPayServer/BTCPayServer.csproj
+    dotnet restore --source $TMP/nuget ${optionalString altcoinSupport ''/p:Configuration="Altcoins-Release" /p:RazorCompileOnBuild=false''} BTCPayServer/BTCPayServer.csproj
```
Why is there a need for `RazorCompileOnBuild=false`? Can you add comment into the source code why is that needed?

nixbitcoin replied:
Actually not needed

prusnak commented on pkgs/applications/blockchains/btcpayserver/update.sh:
```diff
@@ -8,4 +8,4 @@ echo "Updating nbxplorer"
 echo
 echo "Updating btcpayserver"
-../nbxplorer/util/update-common.sh btcpayserver deps.nix
+../nbxplorer/util/update-common.sh btcpayserver deps.nix '"/p:Configuration="Altcoins-Release" /p:RazorCompileOnBuild=false"'
```
Same here - why is `RazorCompileOnBuild=false` needed?

nixbitcoin replied:
Needed, see https://github.com/NixOS/nixpkgs/pull/130162/commits/d17ebc3252fa522ee9e52e61b29501c5f231c32e

nixbitcoin commented:
Very interesting, I'm getting `/build/source/BTCPayServer/BTCPayServer.csproj : error NU1101: Unable to find package Microsoft.AspNetCore.Mvc.Razor.RuntimeCompilation. No packages exist with this id in source(s): tmpsrc` when running with a `deps.nix` generated by `update.sh` without `/p:RazorCompileOnBuild=false`, see for yourself with [this test](https://gist.github.com/nixbitcoin/2886b40c05d79173f0c441385429370c). `/p:RazorCompileOnBuild=false` seems unnecessary in the `default.nix` though, see for yourself with [this test](https://gist.github.com/nixbitcoin/734cd5e23049f7aaa9296bd079bff1c5). Removed there and added comment.

nixbitcoin commented:
> Very interesting, I'm getting `/build/source/BTCPayServer/BTCPayServer.csproj : error NU1101: Unable to find package Microsoft.AspNetCore.Mvc.Razor.RuntimeCompilation. No packages exist with this id in source(s): tmpsrc` when running with a `deps.nix` generated by `update.sh` without `/p:RazorCompileOnBuild=false`, see for yourself with [this test](https://gist.github.com/nixbitcoin/2886b40c05d79173f0c441385429370c).~`/p:RazorCompileOnBuild=false` seems unnecessary in the `default.nix` though, see for yourself with [this test](https://gist.github.com/nixbitcoin/734cd5e23049f7aaa9296bd079bff1c5). Removed there and added comment.~
> 
> EDIT: Needed in both [test](https://gist.github.com/nixbitcoin/2886b40c05d79173f0c441385429370c) failing with `fail: Configuration: This build of BTCPay Server does not support altcoins`

Actually, now I'm not sure. The error might be due to the test script. Running fine with the nix-bitcoin test suite based on https://github.com/nixbitcoin/nixpkgs/commit/50fc1b2b2bcb3438431a2c3c03b6c29e00e940db without `RazorCompileOnBuild=false` in the `default.nix`.

SuperSandro2000 commented on pkgs/applications/blockchains/btcpayserver/default.nix:
```diff
@@ -1,7 +1,9 @@
+  dotnetPackages, dotnetCorePackages, altcoinSupport ? false
 }:
 
+with lib;
+
```
```suggestion
```

nixbitcoin replied:
Fixed in bf6cac3e7a94de1a5dc0da81899ec489780210a7

SuperSandro2000 commented on pkgs/applications/blockchains/btcpayserver/default.nix:
```diff
@@ -34,8 +36,9 @@ stdenv.mkDerivation rec {
-    dotnet restore --source $TMP/nuget BTCPayServer/BTCPayServer.csproj
-    dotnet publish --no-restore --output $out/share/$pname -c Release BTCPayServer/BTCPayServer.csproj
+    # RazorCompileOnBuild needed for altcoin support
+    dotnet restore --source $TMP/nuget ${optionalString altcoinSupport ''/p:Configuration="Altcoins-Release" /p:RazorCompileOnBuild=false''} BTCPayServer/BTCPayServer.csproj
+    dotnet publish --no-restore --output $out/share/$pname ${optionalString altcoinSupport "-c Altcoins-Release"} BTCPayServer/BTCPayServer.csproj
```
```suggestion
    dotnet restore --source $TMP/nuget ${lib.optionalString altcoinSupport ''/p:Configuration="Altcoins-Release" /p:RazorCompileOnBuild=false''} BTCPayServer/BTCPayServer.csproj
    dotnet publish --no-restore --output $out/share/$pname ${lib.optionalString altcoinSupport "-c Altcoins-Release"} BTCPayServer/BTCPayServer.csproj
```

nixbitcoin replied:
Fixed in bf6cac3e7a94de1a5dc0da81899ec489780210a7
